### PR TITLE
[Merged by Bors] - feat: several results on permutations, cycles, etc

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
These are basic results on permutations which will be useful for PR #17522 and  #17047 which
study the centralizer of a permutation. 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/Perm/Centralizer.lean [hunk 1]
```diff
@@ -0,0 +1,1181 @@
+/-
+Copyright (c) 2023 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+
+-/
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Perm/Centralizer.lean:6]:
```suggestion
Authors: Antoine Chambert-Loir
-/
```

File: Mathlib/GroupTheory/Perm/Centralizer.lean [hunk 2]
```diff
@@ -0,0 +1,1181 @@
...
+
+import Mathlib.GroupTheory.NoncommCoprod
+import Mathlib.GroupTheory.NoncommPiCoprod
+import Mathlib.Algebra.Order.BigOperators.Ring.Finset
+import Mathlib.Algebra.Order.BigOperators.GroupWithZero.Multiset
+-- import Mathlib.GroupTheory.GroupAction.SubMulAction
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Perm/Centralizer.lean:12]:
```suggestion
```

File: Mathlib/GroupTheory/Perm/Centralizer.lean [hunk 3]
```diff
@@ -0,0 +1,1181 @@
...
+lemma MonoidHom.commute_noncommPiCoprod
+    {ι : Type*} [Fintype ι] {H : ι → Type*} [∀ i, Monoid (H i)] {P : Type*} [Monoid P]
+    (f : (i : ι) → H i →* P) (comm) (p : P) (hcomm : ∀ i (x : H i), Commute p (f i x))
+    (h : (i : ι) → H i) :
+    Commute p (MonoidHom.noncommPiCoprod f comm h) := by
+  dsimp only [MonoidHom.noncommPiCoprod, MonoidHom.coe_mk, OneHom.coe_mk]
+  apply Finset.noncommProd_induction
+  exact fun x y ↦ Commute.mul_right
+  exact Commute.one_right _
+  exact fun x _ ↦ hcomm x (h x)
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Perm/Centralizer.lean:114]:
```suggestion
  · exact fun x y ↦ Commute.mul_right
  · exact Commute.one_right _
  · exact fun x _ ↦ hcomm x (h x)
```

File: Mathlib/GroupTheory/Perm/Cycle/Factors.lean [hunk 4]
```diff
@@ -128,6 +128,7 @@ theorem cycleOf_eq_one_iff (f : Perm α) [DecidableRel f.SameCycle] : cycleOf f
   · rw [hy, ite_self]
   · exact if_neg (mt SameCycle.apply_eq_self_iff (by tauto))
 
+
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Perm/Cycle/Factors.lean:131]: Please revert spurious changes in other files

File: Mathlib/GroupTheory/Perm/Centralizer.lean [hunk 5]
```diff
@@ -0,0 +1,1181 @@
...
+  rw [eq_comm]
+  exact Iff.symm mul_inv_eq_iff_eq_mul
+
+theorem Subgroup.nat_card_centralizer_nat_card_stabilizer :
+    Nat.card (Subgroup.centralizer {g}) =
+      Nat.card (MulAction.stabilizer (ConjAct G) g) := by
+  simp only [← SetLike.coe_sort_coe, Set.Nat.card_coe_set_eq]
+  rw [Subgroup.centralizer_eq_comap_stabilizer, Subgroup.coe_comap,
+    MulEquiv.toMonoidHom_eq_coe, MonoidHom.coe_coe]
+  erw [Set.preimage_equiv_eq_image_symm]
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Perm/Centralizer.lean:169]: Please check why you need `erw`

File: Mathlib/GroupTheory/Perm/Centralizer.lean [hunk 6]
```diff
@@ -0,0 +1,1181 @@
...
+lemma Finset.disjoint_of_le_iff_left_eq_empty {α : Type*} {u v : Set α} (h : u ⊆ v) :
+    Disjoint u v ↔ u = ∅ := by
+  simp only [disjoint_iff, inf_eq_left.mpr h, Set.bot_eq_empty]
+
+lemma disjoint_of_le_iff_left_eq_bot
+    {α : Type*} [SemilatticeInf α] [OrderBot α] {a b : α} (h : a ≤ b) :
+    Disjoint a b ↔ a = ⊥ := by
+  simp only [disjoint_iff, inf_eq_left.mpr h]
+
+end Disjoint
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Perm/Centralizer.lean:196]: Can these not go earlier? Actually, same question about much more of this file

File: Mathlib/GroupTheory/Perm/Centralizer.lean [hunk 7]
```diff
@@ -0,0 +1,1181 @@
...
+  induction s using Finset.induction_on with
+  | empty => simp
+  | @insert i s hi hrec =>
+    have hs' : (s : Set ι).Pairwise fun i j ↦ Disjoint (k i) (k j) :=
+      hs.mono (by simp only [Finset.coe_insert, Set.subset_insert])
+    rw [Finset.noncommProd_insert_of_not_mem _ _ _ _ hi]
+    apply Equiv.Perm.Disjoint.mul_right (hf i _) (hrec hs' _)
+    · simp
+    · intro j hj
+      exact hf j (Finset.mem_insert_of_mem hj) -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Perm/Centralizer.lean:311]: Remove

**Review by Ruben-VandeVelde — CHANGES_REQUESTED** [review]
Did you mean to submit this for review already?

**Comment by Ruben-VandeVelde** [comment]
And please split this up

File: Mathlib/GroupTheory/Perm/Centralizer.lean (hunk 6 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Centralizer.lean]: I dispatched these lemmas in Data/Set/Basic and Data/Finset/Basic and Order/Disjoint.
For the rest, I'll have a look at what can be done.

Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Centralizer.lean]: @Ruben-VandeVelde : unfortunately, adding 4 more lines to Data/Finset/Basic made it be larger than the 3100-line limit.

File: Mathlib/GroupTheory/Perm/Centralizer.lean (hunk 5 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Centralizer.lean:167]: I don't understand *why*. The new code has 3 other approaches, one works, the two other dont'.

Comment (ADedecker) [reply Mathlib/GroupTheory/Perm/Centralizer.lean:109]: The problem is the coercion from `MulEquiv` to normal functions does not go through `Equiv` (at least not synctactically), so `rw` sees no `Equiv` in the goal. It works if you do `  rw [← EquivLike.coe_coe, Set.preimage_equiv_eq_image_symm]`.

There seems to be some API lacking around cardinal of preimages. In the meantime, I think the following defeq-abuse is the nicest proof we can do:
```suggestion
  rw [Subgroup.centralizer_eq_comap_stabilizer]
  rfl
```

**Comment by AntoineChambert-Loir** [comment]
I have removed the main new part to another PR (#17522). The present PR consists of basic useful lemmas.

**Comment by AntoineChambert-Loir** [comment]
I don't understand what the `large-import` tag means, because it doesn't seem I have modified `import`s.

File: Mathlib/GroupTheory/GroupAction/ConjAct.lean [hunk 8]
```diff
@@ -259,6 +259,14 @@ theorem stabilizer_eq_centralizer (g : G) :
   le_antisymm (le_centralizer_iff.mp (zpowers_le.mpr fun _ => mul_inv_eq_iff_eq_mul.mp)) fun _ h =>
     mul_inv_eq_of_eq_mul (h g (mem_zpowers g)).symm
 
+theorem _root_.Subgroup.centralizer_eq_comap_stabilizer (g : G) :
+    Subgroup.centralizer {g} = Subgroup.comap ConjAct.toConjAct.toMonoidHom
+      (MulAction.stabilizer (ConjAct G) g) := by
+  ext k
+  simp only [mem_centralizer_iff, Set.mem_singleton_iff, forall_eq, ConjAct.toConjAct_smul]
+  rw [eq_comm]
```
Comment (riccardobrasca) [inline Mathlib/GroupTheory/GroupAction/ConjAct.lean:267]: Can you check (quickly) if the need for `rw [eq_comm]` is caused by a lemma written in a way different wrt to the others?

File: Mathlib/GroupTheory/NoncommCoprod.lean [hunk 9]
```diff
@@ -46,6 +46,13 @@ def noncommCoprod (comm : ∀ m n, Commute (f m) (g n)) : M × N →ₙ* P where
   toFun mn := f mn.fst * g mn.snd
   map_mul' mn mn' := by simpa using (comm _ _).mul_mul_mul_comm _ _
 
+/-- Variant of `MulHom.noncommCoprod_apply` with the product written in the other direction` -/
+@[to_additive
+  "Variant of `AddHom.noncommCoprod_apply`, with the product written in the other direction"]
```
Comment (riccardobrasca) [inline Mathlib/GroupTheory/NoncommCoprod.lean:51]:
```suggestion
  "Variant of `AddHom.noncommCoprod_apply`, with the sum written in the other direction"]
```
But maybe I am misunderstanding the statement.

File: Mathlib/GroupTheory/NoncommCoprod.lean [hunk 10]
```diff
@@ -74,6 +81,13 @@ def noncommCoprod : M × N →* P where
   map_one' := by simp only [Prod.fst_one, Prod.snd_one, map_one, mul_one]
   __ := f.toMulHom.noncommCoprod g.toMulHom comm
 
+/-- Variant of `MonoidHom.noncomCoprod_apply` with the product written in the other direction` -/
+@[to_additive
+  "Variant of `AddMonoidHom.noncomCoprod_apply` with the product written in the other direction"]
```
Comment (riccardobrasca) [inline Mathlib/GroupTheory/NoncommCoprod.lean:86]: Same.

File: Mathlib/GroupTheory/Perm/Cycle/Factors.lean [hunk 11]
```diff
@@ -512,6 +512,23 @@ theorem mem_cycleFactorsFinset_support_le {p f : Perm α} (h : p ∈ cycleFactor
...
+  simp only [← hm]
+  exact le_trans (support_zpow_le _ _) (mem_cycleFactorsFinset_support_le c.prop)
+
+theorem mem_support_iff_mem_support_of_mem_cycleFactorsFinset {g : Equiv.Perm α} {x : α} :
+    x ∈ g.support ↔ ∃ c ∈ g.cycleFactorsFinset, x ∈ c.support := by
+  constructor
+  · intro h
+    use g.cycleOf x, cycleOf_mem_cycleFactorsFinset_iff.mpr h
+    rw [mem_support_cycleOf_iff]
+    refine ⟨SameCycle.refl g x, h⟩
```
Comment (riccardobrasca) [inline Mathlib/GroupTheory/Perm/Cycle/Factors.lean:528]:
```suggestion
    exact ⟨SameCycle.refl g x, h⟩
```

File: Mathlib/GroupTheory/Perm/Cycle/Type.lean [hunk 12]
```diff
@@ -69,6 +69,23 @@ theorem cycleType_eq {σ : Perm α} (l : List (Perm α)) (h0 : l.prod = σ)
...
+  -- work on the LHS
+  rw [cycleType, Multiset.count_eq_card_filter_eq]
+  -- rewrite the `Fintype.card` as a `Finset.card`
+  rw [Fintype.subtype_card, Finset.univ_eq_attach, Finset.filter_attach',
+    Finset.card_map, Finset.card_attach]
+  simp only [Function.comp_apply, Finset.card, Finset.filter_val,
+    Multiset.filter_map, Multiset.card_map]
+  apply congr_arg
+  ext c
+  apply congr_arg₂ _ rfl
```
Comment (riccardobrasca) [inline Mathlib/GroupTheory/Perm/Cycle/Type.lean:84]: Can you use `congr` here? Maybe even above.

**Review by riccardobrasca — COMMENTED** [review]
Thanks!

bors d+

File: Mathlib/GroupTheory/GroupAction/ConjAct.lean (hunk 8 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/ConjAct.lean:267]: By `mem_centralizer_iff`, k belongs to the centralizer of {g} means g * k = k * g, while k belongs to the stabilizer of `g` by the conjugation action means that `k * g * k^{-1} = g`.  
It probably makes sense to rephrase `mem_centralizer_iff` in the opposite direction. (But please let
me do that elsewhere!)

File: Mathlib/GroupTheory/NoncommCoprod.lean (hunk 9 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/NoncommCoprod.lean:51]: No, you're right.

File: Mathlib/GroupTheory/GroupAction/ConjAct.lean (hunk 8 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/ConjAct.lean:267]: I let a note that maybe we should do that sometime.

**Pull request closed without merging**